src/high-level/matrix.lisp

```diff
@@ -0,0 +1,560 @@
+    (declare (ignore matrix))
+    (error "LU is not defined for the generic matrix type.")))
+
+;; TODO: Make this one generic and move to lapack-macros
+;;       This is being blocked by the ZUNCSD shenanigans
+(defgeneric csd (matrix p q)
+  (:documentation "Find the Cosine-Sine Decomposition of a matrix X given that it is to be partitioned with upper left block of dimension P-by-Q. Returns the CSD elements (VALUES U SIGMA VT) such that X=U*SIGMA*VT.")
+  (:method ((matrix matrix) p q)
+    (labels ((csd-from-blocks (u1 u2 v1t v2t theta)
+               "Calculates the matrices U, SIGMA, and VT of the CSD of a matrix from its intermediate representation, as calculated from LAPACK-CSD."
+               (let ((p (nrows u1))
+                     (q (nrows v1t))
+                     (m (+ (nrows u1) (nrows u2)))
+                     (r (length theta)))
+                 (let ((u (direct-sum u1 u2))
+                       (sigma (const 0 (list m m) :type (element-type matrix)))
+                       (vt (direct-sum v1t v2t)))
+                   (let ((diag11 (min p q))
+                         (diag12 (min p (- m q)))
+                         (diag21 (min (- m p) q))
+                         (diag22 (min (- m p) (- m q))))
+                     (let ((iden11 (- diag11 r))
+                           (iden12 (- diag12 r))
+                           (iden21 (- diag21 r))
+                           (iden22 (- diag22 r)))
+                       ;; Construct sigma from theta
+                       (loop :for i :from 0 :to (1- iden11)
+                             do (setf (tref sigma i i) 1))
+                       (loop :for i :from iden11 :to (1- diag11)
+                             do (setf (tref sigma i i) (cos (nth (- i iden11) theta))))
+                       (loop :for i :from 0 :to (1- iden12)
+                             do (setf (tref sigma (- p 1 i) (- m 1 i)) -1))
+                       (loop :for i :from iden12 :to (1- diag12)
+                             do (setf (tref sigma (- p 1 i) (- m 1 i))
+                                      (- (sin (nth (- r 1 (- i iden12)) theta)))))
+                       (loop :for i :from 0 :to (1- iden21)
+                             do (setf (tref sigma (- m 1 i) (- q 1 i)) 1))
+                       (loop :for i :from iden21 :to (1- diag21)
+                             do (setf (tref sigma (- m 1 i) (- q 1 i))
+                                      (sin (nth (- r 1 (- i iden21)) theta))))
+                       (loop :for i :from 0 :to (1- iden22)
+                             do (setf (tref sigma (+ p i) (+ q i)) 1))
+                       (loop :for i :from iden22 :to (1- diag22)
+                             do (setf (tref sigma (+ p i) (+ q i)) (cos (nth (- i iden22) theta))))))
+                   (values u sigma vt)))))
+      (multiple-value-bind (u1 u2 v1t v2t theta) (lapack-csd matrix p q)
+        (csd-from-blocks u1 u2 v1t v2t theta)))))
+
+(defgeneric inv (matrix)
```

Comment:
This seems to return a value even for non-invertible matrices, e.g.
```
MAGICL> (inv (ones '(2 2)))
#<MATRIX/DOUBLE-FLOAT (2x2):
   1.000     1.000
   1.000     0.000>
```
This should instead signal a condition. As a side note, there are meaningful generalizations of the matrix inverse (e.g. Moore-Penrose pseudoinverse), but I don't recognize the above as being one of those either.

Reply:
Starting to look into this. This matrix is clearly singular, which should cause the `dgetrf` LAPACK routine (lu factor) to return a negative `INFO` value, however it seems to be returning zero no matter what the input is.

Reply:
I think it might be due to how the current LAPACK bindings are written. Consider
```
(COMMON-LISP:DEFUN %DGETRF (M N A LDA IPIV INFO)
  (COMMON-LISP:DECLARE (COMMON-LISP:INLINE %%DGETRF)
                       (COMMON-LISP:TYPE (COMMON-LISP:SIGNED-BYTE 32) M)
                       (COMMON-LISP:TYPE (COMMON-LISP:SIGNED-BYTE 32) N)
                       (COMMON-LISP:TYPE
                        (COMMON-LISP:SIMPLE-ARRAY COMMON-LISP:DOUBLE-FLOAT) A)
                       (COMMON-LISP:TYPE (COMMON-LISP:SIGNED-BYTE 32) LDA)
                       (COMMON-LISP:TYPE
                        (COMMON-LISP:SIMPLE-ARRAY (COMMON-LISP:SIGNED-BYTE 32)
                         (COMMON-LISP:*))
                        IPIV)
                       (COMMON-LISP:TYPE (COMMON-LISP:SIGNED-BYTE 32) INFO))
  (CFFI:WITH-FOREIGN-OBJECTS ((M-REF6702 ':INT32) (N-REF6703 ':INT32)
                              (LDA-REF6705 ':INT32) (INFO-REF6707 ':INT32))
    (COMMON-LISP:SETF (CFFI:MEM-REF M-REF6702 :INT32) M)
    (COMMON-LISP:SETF (CFFI:MEM-REF N-REF6703 :INT32) N)
    (COMMON-LISP:SETF (CFFI:MEM-REF LDA-REF6705 :INT32) LDA)
    (COMMON-LISP:SETF (CFFI:MEM-REF INFO-REF6707 :INT32) INFO)
    (MAGICL.CFFI-TYPES:WITH-ARRAY-POINTERS ((A-REF6704 A) (IPIV-REF6706 IPIV))
      (%%DGETRF M-REF6702 N-REF6703 A-REF6704 LDA-REF6705 IPIV-REF6706
                INFO-REF6707))))
```
(from https://github.com/rigetti/magicl/blob/feature/magicl-ng/src/bindings/lapack02-cffi.lisp#L3725).
`info` is just a number. mutations to it within the `defun` body will not be visible to callers.

Any workaround is going to require rebuilding the lapack bindings. 
- One option is to change them to explicitly take boxed values by default (say, a length 1 vector), and then users need to manage boxing/unboxing around the call site. (barf)
- Another option is to change the bindings so that the functions return the `cffi:mem-ref` values corresponding to the defun arguments. This is kind of heavyweight but shouldn't break existing code. Then your `lapack-inv` routine can just check the last of the return values (corresponding to `info`). 

There are probably other approaches.

Reply:
Added check for singular matrices in 33973d5 and fixing bindings can be put into a future PR

Reply:
Created #84